### PR TITLE
[DeepSeek-V4] Support BF16 KV Cache Decode Path

### PR DESCRIPTION
**Pull request by yuanlehome**
## Motivation

FlashMLA
- https://github.com/deepseek-ai/FlashMLA/pull/186
- https://github.com/sgl-project/FlashMLA/pull/7

When serving DeepSeek-V4 with `--kv-cache-dtype bf16`, the entire decode pipeline needs to support BF16 KV cache storage, write, and attention computation. Previously, the DSV4 sparse decode path only supported FP8 quantized KV cache. This PR adds end-to-end BF16 KV cache support, including memory pool layout adaptation, KV write kernels, and FlashMLA decode kernel invocation.

BF16 mode is suitable for accuracy-sensitive workloads (e.g., reasoning tasks), avoiding quantization error inherent in FP8.

### Memory Pool (`deepseek_v4_memory_pool.py`)
- Differentiate BF16/FP8 layout based on `kv_cache_dtype`: BF16 uses 1024 bytes/token (448×2 + 64×2), FP8 uses 584 bytes/token
- Remove unnecessary 576-byte page padding in BF16 mode
- `set_key_buffer` dispatches on `NopeBf16RopeBf16Pack` type
- `get_key_buffer` uniformly returns raw uint8 buffer; downstream kernels view as needed

### Index Buffer Accessor (`dsv4/index_buf_accessor.py`)
- Added `NopeBf16RopeBf16Pack` dataclass (nope bf16 + rope bf16, no scale)
- Added `SetKBf16` class with corresponding Triton kernel to write bf16 nope/rope directly into page buffer

### Attention Backend (`deepseek_v4_backend.py`)
- Added `is_bf16_kv_cache` flag based on `server_args.kv_cache_dtype`
- SWA key store: BF16 mode skips FP8 quantization, uses `split_nope_bf16_rope_bf16`
- Sparse decode: `.view(torch.bfloat16)` on uint8 buffer before FlashMLA call
- Changed `import sgl_kernel.flash_mla` to `import flash_mla` (standalone FlashMLA package)

### Quant K Cache (`dsv4/quant_k_cache.py`)
- Added `split_nope_bf16_rope_bf16`: directly splits nope/rope by dimension without quantization in BF16 mode

### JIT Kernels (`jit_kernel/dsv4/`)
- Added `fused_k_norm_rope_flashmla_bf16` and `compress_norm_rope_store_bf16`
- Added corresponding CUDA kernels (`fused_norm_rope_v2.cuh`, `main_norm_rope.cuh`)

## Accuracy Tests

- Compared generation outputs between `--kv-cache-dtype fp8` and `--kv-cache-dtype bf16` using DeepSeek-V4-Flash-FP8 model
- BF16 mode logits align with PyTorch reference attention implementation

sgl-project/DeepSeek-V4-Flash-FP8: 88.19%
image
image

```shell
# Accuracy: 0.952
# Invalid: 0.000
# Latency: 102.807 s
# Output throughput: 1148.554 token/s
python bench_sglang.py --host http://localhost  --port 30100 --data-path /data00 --num-questions 5000 --parallel 100


# Repeat: 8, mean: 0.922
# Scores: ['0.938', '0.938', '0.906', '0.906', '0.938', '0.906', '0.906', '0.938']
# Mean latency: 1082.458 s
python -m sglang.test.run_eval --port 30100 --eval-name gpqa --num-examples 32 --max-tokens 128000 --repeat 8 --top-p 0.95 --temperature 1.0 --thinking-mode deepseek-v3
```

## Speed Tests and Profiling

```shell
# test on B200
sglang serve \
  --trust-remote-code \
  --model-path sgl-project/DeepSeek-V4-Flash-FP8 \
  --tp 8 \
  --dp 8 \
  --ep 8 \
  --enable-dp-attention \
  --moe-a2a-backend deepep \
  --deepep-config '{"normal_dispatch":{"num_sms":96},"normal_combine":{"num_sms":96}}' \
  --host 0.0.0.0 \
  --port 30100 \
  --nccl-port 18201 \
  --tool-call-parser deepseekv4 \
  --reasoning-parser deepseek-v4 \
  --kv-cache-dtype bf16 \
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29000887943](https://github.com/sgl-project/sglang/actions/runs/29000887943)
Latest PR Test (Extra): :x: [Run #29000887762](https://github.com/sgl-project/sglang/actions/runs/29000887762)

### CODE REVIEW

File: sgl-kernel/cmake/flashmla.cmake [hunk 1]
```diff
@@ -144,6 +145,10 @@ if(FLASHMLA_ENABLE_SM100)
         ${repo-flashmla_SOURCE_DIR}/csrc/sm100/decode/head64/instantiations/v32.cu
         ${repo-flashmla_SOURCE_DIR}/csrc/sm100/decode/head64/instantiations/model1.cu
         ${repo-flashmla_SOURCE_DIR}/csrc/sm100/prefill/sparse/fwd_for_small_topk/head128/instantiations/phase1_decode_k512.cu
+
+        # sm100 sparse decode bf16.
+        ${repo-flashmla_SOURCE_DIR}/csrc/sm100/decode/head64_bf16/instantiations/v32.cu
+        ${repo-flashmla_SOURCE_DIR}/csrc/sm100/decode/head64_bf16/instantiations/model1.cu
```
Comment (zeroRains) [inline sgl-kernel/cmake/flashmla.cmake:151]: file in https://github.com/sgl-project/FlashMLA/pull/7
